.rubocop.yml

```diff
@@ -1,20 +1,44 @@
 Rails/CreateTableWithTimestamps:
   Enabled: false
 Rails/HasManyOrHasOneDependent:
-  Enabled: false
+  Enabled: true
+# if a codebase is on an older version on rails override this in the local rubocop.yml
 Rails/HttpPositionalArguments:
-  Enabled: false
+  Enabled: true
 Rails/LexicallyScopedActionFilter:
   Enabled: false
 Rails/Presence:
-  Enabled: false
+  Enabled: true
 Rails/FilePath:
   EnforcedStyle: arguments
-# We should enable this later but there is currently a bug in rubocop which
-# breaks this
-# Rails/RequestReferer:
-#   EnforcedStyle: referrer
-#
+Rails/RequestReferer:
+  EnforcedStyle: referrer
+# Rails generators break this rule out of the box
+Style/ClassAndModuleChildren:
+  Enabled: false
```

Comment:
this is one are where I disagree with rails. I think that not requiring the nested structure introduces a lot of constant namespacing issues and is something you can run into in rails fairly often with namespaced models: 
```ruby
module Admin
  FOO = 'foo'
  def self.table_name_prefix
    'admin_'
  end
end

# Lookup => [Admin::User, User]
module Admin
class User < ApplicationRecord
  def self.bar
    puts FOO #=> 'foo'
  end
end
end

# Lookup => [Admin::User]
class Admin::User < ApplicationRecord
  def self.bar
    puts FOO #=> uninitialized constant Admin::User::FOO
  end

  def self.foo_bar
    puts Admin::FOO #=> uninitialized constant Admin::User::FOO
  end
end

```

Reply:
I agree with you, but how do we handle it, should we update our generator templates to use the nested module/class?

Reply:
let me see if I can whip up something for the generators tonight

Reply:
Just something I whipped up, needs to be cleaned up, but it should be fairly generic.
Late night code ahead:
```ruby
# model generator
require "rails/generators/active_record/model/model_generator"

class ModelGenerator < ActiveRecord::Generators::ModelGenerator
  private

  def nested_namespacing(class_path, &block)
    content = capture(&block)
    class_path.each do |module_namespace|
      content = wrap_with_module_namespace(module_namespace, content)
    end
    concat(content)
  end

  def wrap_with_module_namespace(module_namespace, content)
    content = indent(content).chomp
    "module #{module_namespace.camelize}\n#{content}\nend\n"
  end
end

# Model template

<% module_namespacing do -%>
<% nested_namespacing(class_path) do -%>
class <%= name.split('/').last.camelize %> < <%= parent_class_name.classify %>
<% attributes.select(&:reference?).each do |attribute| -%>
  belongs_to :<%= attribute.name %><%= ', polymorphic: true' if attribute.polymorphic? %><%= ', required: true' if attribute.required? %>
<% end -%>
<% attributes.select(&:token?).each do |attribute| -%>
  has_secure_token<% if attribute.name != "token" %> :<%= attribute.name %><% end %>
<% end -%>
<% if attributes.any?(&:password_digest?) -%>
  has_secure_password
<% end -%>
end
<% end -%>
<% end -%>

# Module Template

<% module_namespacing do -%>
<% nested_namespacing(class_path[0...-1]) do %>
module <%= class_path.pop.camelize %>
  def self.table_name_prefix
    '<%= namespaced? ? namespaced_class_path.join('_') : class_path.join('_') %>_'
  end
end
<% end -%>
<% end -%>
```

Reply:
example is only for models but we could probably extract that out, still need to break stuff out into helpers (stuff like the new class name), and it would require overwriting quite a few templates. so there's some caveats